File: sharehodl-ios/ShareHODL/Models/Chain.swift
import Foundation

/// Supported blockchain networks
enum Chain: String, CaseIterable, Identifiable, Codable {
    case sharehodl = "sharehodl"
    case bitcoin = "bitcoin"
    case litecoin = "litecoin"
    case ethereum = "ethereum"

    var id: String { rawValue }

    /// Display name
    var name: String {
        switch self {
        case .sharehodl: return "ShareHODL"
        case .bitcoin: return "Bitcoin"
        case .litecoin: return "Litecoin"
        case .ethereum: return "Ethereum"
        }
    }

    /// Token symbol
    var symbol: String {
        switch self {
        case .sharehodl: return "HODL"
        case .bitcoin: return "BTC"
        case .litecoin: return "LTC"
        case .ethereum: return "ETH"
        }
    }

    /// BIP44 coin type for key derivation
    /// https://github.com/satoshilabs/slips/blob/master/slip-0044.md
    var coinType: UInt32 {
        switch self {
        case .sharehodl: return 118  // Cosmos
        case .bitcoin: return 0
        case .litecoin: return 2
        case .ethereum: return 60
        }
    }

    /// Address prefix for Bech32 encoding (Cosmos-style chains)
    var bech32Prefix: String? {
        switch self {
        case .sharehodl: return "hodl"
        case .bitcoin: return "bc"      // Native SegWit
        case .litecoin: return "ltc"    // Native SegWit
        case .ethereum: return nil      // Ethereum uses hex addresses
        }
    }

    /// Whether this chain uses Cosmos-style addresses
    var isCosmosStyle: Bool {
        switch self {
        case .sharehodl: return true
        default: return false
        }
    }

    /// Whether this chain uses Bitcoin-style SegWit addresses
    var isBitcoinStyle: Bool {
        switch self {
        case .bitcoin, .litecoin: return true
        default: return false
        }
    }

    /// Whether this chain uses Ethereum-style addresses
    var isEthereumStyle: Bool {
        switch self {
        case .ethereum: return true
        default: return false
        }
    }

    /// Number of decimal places for display
    var decimals: Int {
        switch self {
        case .sharehodl: return 6   // uhodl
        case .bitcoin: return 8     // satoshis
        case .litecoin: return 8    // litoshis
        case .ethereum: return 18   // wei
        }
    }

    /// Smallest unit name
    var smallestUnit: String {
        switch self {
        case .sharehodl: return "uhodl"
        case .bitcoin: return "satoshi"
        case .litecoin: return "litoshi"
        case .ethereum: return "wei"
        }
    }

    /// Chain icon (SF Symbol name)
    var iconName: String {
        switch self {
        case .sharehodl: return "h.circle.fill"
        case .bitcoin: return "bitcoinsign.circle.fill"
        case .litecoin: return "l.circle.fill"
        case .ethereum: return "e.circle.fill"
        }
    }

    /// Chain color for UI
    var colorHex: String {
        switch self {
        case .sharehodl: return "#6366F1"  // Indigo
        case .bitcoin: return "#F7931A"    // Bitcoin orange
        case .litecoin: return "#345D9D"   // Litecoin blue
        case .ethereum: return "#627EEA"   // Ethereum purple
        }
    }

    /// Explorer URL template (replace {address} or {tx})
    var explorerAddressURL: String {
        switch self {
        case .sharehodl: return "https://explorer.sharehodl.com/address/{address}"
        case .bitcoin: return "https://mempool.space/address/{address}"
        case .litecoin: return "https://blockchair.com/litecoin/address/{address}"
        case .ethereum: return "https://etherscan.io/address/{address}"
        }
    }

    var explorerTxURL: String {
        switch self {
        case .sharehodl: return "https://explorer.sharehodl.com/tx/{tx}"
        case .bitcoin: return "https://mempool.space/tx/{tx}"
        case .litecoin: return "https://blockchair.com/litecoin/transaction/{tx}"
        case .ethereum: return "https://etherscan.io/tx/{tx}"
        }
    }
}

/// Represents a wallet account on a specific chain
struct ChainAccount: Identifiable, Codable {
    let id: UUID
    let chain: Chain
    let address: String
    let derivationPath: String
    var balance: String
    var balanceUSD: String?

    init(chain: Chain, address: String, derivationPath: String, balance: String = "0", balanceUSD: String? = nil) {
        self.id = UUID()
        self.chain = chain
        self.address = address
        self.derivationPath = derivationPath
        self.balance = balance
        self.balanceUSD = balanceUSD
    }

    /// Formatted balance with symbol
    var formattedBalance: String {
        "\(balance) \(chain.symbol)"
    }

    /// Short address for display (first 8...last 6)
    var shortAddress: String {
        guard address.count > 16 else { return address }
        let prefix = String(address.prefix(10))
        let suffix = String(address.suffix(6))
        return "\(prefix)...\(suffix)"
    }

    /// Explorer URL for this address
    var explorerURL: URL? {
        let urlString = chain.explorerAddressURL.replacingOccurrences(of: "{address}", with: address)
        return URL(string: urlString)
    }
}

File: sharehodl-android/app/src/main/java/com/sharehodl/model/Chain.kt
package com.sharehodl.model

import androidx.compose.ui.graphics.Color

/**
 * Supported blockchain networks
 */
enum class Chain(
    val displayName: String,
    val symbol: String,
    val coinType: Int,  // BIP44 coin type
    val bech32Prefix: String?,
    val decimals: Int,
    val smallestUnit: String,
    val colorHex: Long,
    val explorerAddressUrl: String,
    val explorerTxUrl: String,
    val isToken: Boolean = false,  // True for tokens like USDT, USDC
    val parentChain: String? = null  // Parent chain for tokens
) {
    SHAREHODL(
        displayName = "ShareHODL",
        symbol = "HODL",
        coinType = 118,  // Cosmos
        bech32Prefix = "hodl",
        decimals = 6,
        smallestUnit = "uhodl",
        colorHex = 0xFF6366F1,  // Indigo
        explorerAddressUrl = "https://explorer.sharehodl.com/address/{address}",
        explorerTxUrl = "https://explorer.sharehodl.com/tx/{tx}"
    ),
    BITCOIN(
        displayName = "Bitcoin",
        symbol = "BTC",
        coinType = 0,
        bech32Prefix = null,  // Legacy addresses (1..., 3...)
        decimals = 8,
        smallestUnit = "satoshi",
        colorHex = 0xFFF7931A,  // Bitcoin orange
        explorerAddressUrl = "https://mempool.space/address/{address}",
        explorerTxUrl = "https://mempool.space/tx/{tx}"
    ),
    ETHEREUM(
        displayName = "Ethereum",
        symbol = "ETH",
        coinType = 60,
        bech32Prefix = null,  // Ethereum uses 0x hex addresses
        decimals = 18,
        smallestUnit = "wei",
        colorHex = 0xFF627EEA,  // Ethereum purple
        explorerAddressUrl = "https://etherscan.io/address/{address}",
        explorerTxUrl = "https://etherscan.io/tx/{tx}"
    ),
    USDT(
        displayName = "Tether",
        symbol = "USDT",
        coinType = 60,  // ERC-20 on Ethereum
        bech32Prefix = null,
        decimals = 6,
        smallestUnit = "micro",
        colorHex = 0xFF26A17B,  // Tether green
        explorerAddressUrl = "https://etherscan.io/address/{address}",
        explorerTxUrl = "https://etherscan.io/tx/{tx}",
        isToken = true,
        parentChain = "ETH"
    ),
    USDC(
        displayName = "USD Coin",
        symbol = "USDC",
        coinType = 60,  // ERC-20 on Ethereum
        bech32Prefix = null,
        decimals = 6,
        smallestUnit = "micro",
        colorHex = 0xFF2775CA,  // USDC blue
        explorerAddressUrl = "https://etherscan.io/address/{address}",
        explorerTxUrl = "https://etherscan.io/tx/{tx}",
        isToken = true,
        parentChain = "ETH"
    ),
    BNB(
        displayName = "BNB",
        symbol = "BNB",
        coinType = 714,  // BSC
        bech32Prefix = null,
        decimals = 18,
        smallestUnit = "jager",
        colorHex = 0xFFF3BA2F,  // Binance yellow
        explorerAddressUrl = "https://bscscan.com/address/{address}",
        explorerTxUrl = "https://bscscan.com/tx/{tx}"
    ),
    LITECOIN(
        displayName = "Litecoin",
        symbol = "LTC",
        coinType = 2,
        bech32Prefix = null,  // Legacy addresses (L..., M...)
        decimals = 8,
        smallestUnit = "litoshi",
        colorHex = 0xFF345D9D,  // Litecoin blue
        explorerAddressUrl = "https://blockchair.com/litecoin/address/{address}",
        explorerTxUrl = "https://blockchair.com/litecoin/transaction/{tx}"
    ),
    DOGECOIN(
        displayName = "Dogecoin",
        symbol = "DOGE",
        coinType = 3,
        bech32Prefix = null,
        decimals = 8,
        smallestUnit = "koinu",
        colorHex = 0xFFC2A633,  // Doge gold
        explorerAddressUrl = "https://dogechain.info/address/{address}",
        explorerTxUrl = "https://dogechain.info/tx/{tx}"
    ),
    SOLANA(
        displayName = "Solana",
        symbol = "SOL",
        coinType = 501,
        bech32Prefix = null,
        decimals = 9,
        smallestUnit = "lamport",
        colorHex = 0xFF9945FF,  // Solana purple
        explorerAddressUrl = "https://solscan.io/account/{address}",
        explorerTxUrl = "https://solscan.io/tx/{tx}"
    ),
    XRP(
        displayName = "XRP",
        symbol = "XRP",
        coinType = 144,
        bech32Prefix = null,
        decimals = 6,
        smallestUnit = "drop",
        colorHex = 0xFF23292F,  // XRP dark
        explorerAddressUrl = "https://xrpscan.com/account/{address}",
        explorerTxUrl = "https://xrpscan.com/tx/{tx}"
    ),
    TRON(
        displayName = "TRON",
        symbol = "TRX",
        coinType = 195,
        bech32Prefix = null,
        decimals = 6,
        smallestUnit = "sun",
        colorHex = 0xFFFF0013,  // Tron red
        explorerAddressUrl = "https://tronscan.org/#/address/{address}",
        explorerTxUrl = "https://tronscan.org/#/transaction/{tx}"
    ),
    POLYGON(
        displayName = "Polygon",
        symbol = "MATIC",
        coinType = 60,  // Uses Ethereum path
        bech32Prefix = null,
        decimals = 18,
        smallestUnit = "wei",
        colorHex = 0xFF8247E5,  // Polygon purple
        explorerAddressUrl = "https://polygonscan.com/address/{address}",
        explorerTxUrl = "https://polygonscan.com/tx/{tx}"
    ),
    AVALANCHE(
        displayName = "Avalanche",
        symbol = "AVAX",
        coinType = 9000,
        bech32Prefix = null,
        decimals = 18,
        smallestUnit = "nAVAX",
        colorHex = 0xFFE84142,  // Avalanche red
        explorerAddressUrl = "https://snowtrace.io/address/{address}",
        explorerTxUrl = "https://snowtrace.io/tx/{tx}"
    );

    val color: Color get() = Color(colorHex)

    val isCosmosStyle: Boolean get() = this == SHAREHODL
    val isBitcoinStyle: Boolean get() = this == BITCOIN || this == LITECOIN || this == DOGECOIN
    val isEthereumStyle: Boolean get() = this == ETHEREUM || this == BNB || this == POLYGON || this == AVALANCHE || isToken

    fun getExplorerUrl(address: String): String =
        explorerAddressUrl.replace("{address}", address)

    fun getTxExplorerUrl(txHash: String): String =
        explorerTxUrl.replace("{tx}", txHash)

    companion object {
        fun fromCoinType(coinType: Int): Chain? =
            entries.find { it.coinType == coinType }

        // Major chains for the main wallet view
        val majorChains = listOf(SHAREHODL, BITCOIN, ETHEREUM, USDT, USDC, BNB, LITECOIN, DOGECOIN, SOLANA, XRP, TRON, POLYGON, AVALANCHE)
    }
}

/**
 * Represents a wallet account on a specific chain
 */
data class ChainAccount(
    val chain: Chain,
    val address: String,
    val derivationPath: String,
    val balance: String = "0",
    val balanceUsd: String? = null,
    val transactions: List<CryptoTransaction> = emptyList()
) {
    val formattedBalance: String get() = "$balance ${chain.symbol}"

    val shortAddress: String get() {
        if (address.length <= 16) return address
        return "${address.take(10)}...${address.takeLast(6)}"
    }

    val explorerUrl: String get() = chain.getExplorerUrl(address)
}

/**
 * Represents a crypto transaction
 */
data class CryptoTransaction(
    val txHash: String,
    val chain: Chain,
    val type: TransactionType,
    val amount: String,
    val amountUsd: String? = null,
    val fromAddress: String,
    val toAddress: String,
    val timestamp: Long,
    val status: TransactionStatus,
    val fee: String? = null,
    val confirmations: Int = 0
) {
    val formattedAmount: String get() = "$amount ${chain.symbol}"

    val shortTxHash: String get() {
        if (txHash.length <= 16) return txHash
        return "${txHash.take(8)}...${txHash.takeLast(8)}"
    }

    val explorerUrl: String get() = chain.getTxExplorerUrl(txHash)

    val formattedDate: String get() {
        val date = java.util.Date(timestamp)
        val format = java.text.SimpleDateFormat("MMM dd, yyyy HH:mm", java.util.Locale.getDefault())
        return format.format(date)
    }

    val isReceived: Boolean get() = type == TransactionType.RECEIVE
}

enum class TransactionType {
    SEND,
    RECEIVE,
    SWAP,
    CONTRACT_CALL
}

enum class TransactionStatus {
    PENDING,
    CONFIRMED,
    FAILED
}
